### ISSUE TITLE
Tensors that share same underlying storage to also share gradient storage

### ISSUE THREAD

**Issue by thomasw21** [pytorch/pytorch#102337]
### 🚀 The feature, motivation and pitch

Hello Pytorch Team! 

I have a question concerning the way gradients are handled, in particular grad attribute. It's a bit hard to explain but in a very short manner I want "views" of a given parameter to have "synced" gradients with the original parameter (shared memory storage). In short the following script tries to mimick what I want to do.

```python
import torch
from torch import nn

def hack_to_sync_grad(param, *view_args, **view_kwargs):
    with torch.no_grad():
        param_view = param.view(*view_args, **view_kwargs)

    # We need sliced tensor to also get the gradient in order to run optimizer on them
    # TODO @thomasw21: It's really had to make sure that our sliced view keeps the same memory space as the original gradient
    def sync_grad(orig_grad):
        assert orig_grad.is_contiguous()
        if param_view.grad is None:
            # The gradient was reset to None, we need to reset the coalesced_tensor.grad as well
            param.grad = None

        if param.grad is None:
            param_view.grad = orig_grad.view(*view_args, **view_kwargs)
        else:
            assert param_view.grad is not None
            # TODO @thomasw21: How do I check that it's still the same memory allocation?
            # One way is to re-assign the existing pointer
            param_view.grad = param.grad.view(*view_args, **view_kwargs)
        return orig_grad

    param.register_hook(sync_grad)
    return param_view

def main():
    model = nn.Linear(2,3)

    weight_view = model.weight.view(-1) # This fails as it doesn't propagate gradients to this view
    # weight_view = hack_to_sync_grad(model.weight, -1) # This is fairly hacky since it relies on the backward hook to sync

    random_input = torch.randn(5,2)
    model(random_input).sum().backward()

    for name, param in model.named_parameters():
        assert param.grad is not None

    # My goal is such that a view of a parameters should also be able to get the "sharded" gradient
    assert weight_view.grad is not None
    assert weight_view.grad.untyped_storage().data_ptr() == model.weight.grad.untyped_storage().data_ptr()

if __name__ == "__main__":
    main()
```

In this example I'm using `view` but I'd like to support things like slices as well. So very concretely what I'm interested in is customising a grad getter. This would allow one to have a syncing mechanism that doesn't rely on the backward hook for syncing.

Solution found as of now:
 - Creating a new property of a new class:
 ```python
 class ViewTensor(torch.Tensor):
    """We want view tensor"""
    def __new__(cls, data, *view_args, **view_kwargs):
        new_tensor = super().__new__(cls, data.view(*view_args, **view_kwargs))
        new_tensor.orig_data = data
        new_tensor.view_args = view_args
        new_tensor.view_kwargs = view_kwargs
        return new_tensor

    @property
    def grad(self):
        return self.orig_data.grad.view(*self.view_args, **self.view_kwargs)

    @grad.setter
    def grad(self, value):
        self.orig_data.grad = value.view_as(self.orig_data)

    @grad.deleter
    def grad(self):
        del self.orig_data.grad

def use_new_tensor_class(param, *view_args, **view_kwargs):
    with torch.no_grad():
        param_view = ViewTensor(param, *view_args, **view_kwargs)
    return param_view
 ```
 Maybe this solution can generalize by allowing for torch.Tensor to pass grad getter/setter/deleter
 
 cc @albanD with whom I initially discussed

**Comment by awgu** [comment]
Hi @thomasw21! 

Tangential to this issue: I recall seeing that you were interested in this in relation to ZeRO. I was curious what the use case was and why existing solutions are insufficient.

**Comment by thomasw21** [comment]
Right! The reason why I need this is I need to propagate `grad` to the parameters I'm passing to the optimizer. Basically it very roughly looks like that:

```python
model = nn.Linear(3,10, bias=False)

# slice parameter
step = model.weight.numel() // group.size()
sliced_param = model.view(-1)[group.rank() * step: (group.rank() + 1) * step]

# setup optimizer with a tensor that shares storage with original parameters
optimizer = AdamW([sliced_param])
```

Typically `model.weight.grad` gets populated during backward pass, but it never goes "up" to `sliced_param`. There might be a way to hack the auto grad system, by first defining `sliced_param` and computing `model.weight` afterwards, but it really sounds like a system when writing a custom getter allows for syncing with no regards to backward pass (typically if for X reasons someone wants to play with `.grad` attribute.

**Comment by albanD** [comment]
Would something like the following work for you:
We add a new `t.register_grad_hook(hook)` where `hook(self, original_grad) -> new_grad` that will be called when the python `.grad` field is used.
Note that we will need to update the warning in the default getter to make sure it doesn't fire spuriously when a hook is registered.

In this case, you can have
```python
def hook(self, original_grad):
    assert original_grad is None
    # Assume we're a view into a bigger Tensor!
    return self._base.grad.as_strided(self.size(), self.stride(), self.storage_offset())
```

cc @ezyang @soulitzer what do you guys think?

**Comment by albanD** [comment]
If you want setter/deleter, we can do `register_grad_hook(getter_hooks, setter=setter_hook, deleter=deleter_hook)`.
The problem is that the `.grad` field is not actually populated from python by the autograd engine. So the setter_hook might be a bit surprising. Also the engine can do `t.grad += new_grad` which wouldn't trigger any of these.

To avoid these questions, I limited my proposal to just a getter. But if we can flesh this out for all 3 if you want.

**Comment by thomasw21** [comment]
Oh just a question, I'm fine with just a getter.

**Comment by soulitzer** [comment]
The subclass solution doesn't seem too bad for this use case actually. @thomasw21 do you run into any issues using your subclass?

But I suppose .grad hooks would be easier for the next person who wants something similar to discover.

**Comment by soulitzer** [comment]
One issue with your Tensor subclass is that whatever operations you do later on will also have to be done with Tensor subclasses, which can reduce perf in certain cases.

A way to fix that is to use a torch dispatch wrapper subclass, which automatically unwraps when you do a non-view operation. 

<details>

<summary>
click to see example implementation
</summary>

```python
import torch
from torch.utils._pytree import tree_map

class ViewGradTensor(torch.Tensor):
    # Based on torch/testing/_internal/logging_tensor.py
    elem: torch.Tensor

    __slots__ = ['elem', 'base']

    __torch_function__ = torch._C._disabled_torch_function_impl

    @staticmethod
    def __new__(cls, elem, base, *args, **kwargs):
        # The wrapping tensor (ViewGradTensor) shouldn't hold any
        # memory for the class in question, but it should still
        # advertise the same device as before
        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
            cls, elem.size(),
            strides=elem.stride(), storage_offset=elem.storage_offset(),
            # TODO: clone storage aliasing
            dtype=elem.dtype, layout=elem.layout,
            device=elem.device, requires_grad=kwargs.get("requires_grad", False)
        )
        # ...the real tensor is held as an element on the tensor.
        r.elem = elem.detach() if r.requires_grad else elem
        return r

    def __init__(self, elem, base):
        super().__init__()
        self.base = base

    def __repr__(self):
        return super().__repr__(tensor_contents=f"{self.elem}")

    @classmethod
    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
        ret_alias_info = func._schema.returns[0].alias_info
        is_view = ret_alias_info is not None and not ret_alias_info.is_write

        def unwrap(e):
            return e.elem if isinstance(e, cls) else e

        def maybe_wrap(e):
            # Rewrap only if the func is a view
            if isinstance(e, torch.Tensor) and is_view:
                # Assume the first input is aliased
                return cls(e, args[0].base)
            else:
                return e
        return tree_map(maybe_wrap, func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))

    def _get_grad(self):
        return self.base.grad.as_strided(self.size(), self.stride(), self.storage_offset())

    grad = property(_get_grad, None, None)

a = torch.ones(2, 3, requires_grad=True)
a.grad = torch.tensor([[2., 3., 4.], [2., 3., 4.]])

print(a.grad)  # tensor([[2., 3., 4.],
               #         [2., 3., 4.]])
b = ViewGradTensor(a[0], a)
print(b.grad)  # tensor([2., 3., 4.])
print(f"{b._is_view()=}")   # False (this is wrong unfortunately)

# If we do more views, the ViewGradTensor is preserved
c = b[:]
print(c)  # ViewGradTensor([1., 1., 1.], grad_fn=<AliasBackward0>)
print(f"{c._is_view()=}")  # True
print(c.grad) # tensor([2., 3., 4.])

# When we do a non-view op, it is no longer a ViewGradTensor
d = c.clone()
print(d) # tensor([1., 1., 1.])
```

</details>

**Comment by thomasw21** [comment]
With the solution given, is it expected that `data_ptr` method doesn't work on the subclass?
```
>>> print(a.data_ptr(), b.data_ptr(), b.elem.data_ptr())
5744655232 0 5744655232
```
I can always override it, but I was under the impression that `__torch_dispatch__` would magically unwrap my custom class into torch tensor anywhere required?

**Comment by albanD** [comment]
You will have to override this method if you want to. Otherwise data_ptr() is returning the actual storage pointer. Which is nullptr -> 0 here.

**Comment by thomasw21** [comment]
Hum maybe I don't understand why it's not passing through the torch dispatch mechanism? Normal tensor subclassing would've worked normally.

Also do you know which methods are impacted?

**Comment by albanD** [comment]
I don't think we have a list no :/
The methods that are handled are all the native functions (as registered to the dispatcher) and special handling for metada (size, stride, storage offset, etc).
Any other function that is python-only (like numpy conversion for example) or too low level concepts (data_ptr) are not caught.
